File: csharp/modules/Numbers.Utils/Numbers.Utils.Operations/Factorial.Utils/FactorialUtils.cs
using System.Numerics;

namespace Prog.Numbers.Utils.Operations;

public class FactorialUtils
{
    /// <summary> <h1><c>Factorial</c></h1>
    /// <br/><br/>This method is used to find the factorial of a
    /// number.<br/><br/>
    ///
    /// <b>Arguments:</b>
    ///
    /// <list type="bullet">
    /// <item>
    /// <c>num: int</c> - The number for which the factorial is to be found.
    /// </item>
    /// </list>
    ///
    /// <b>Return:</b>
    ///
    /// <list type="bullet">
    /// <item>
    /// <c>BigInteger</c> - The factorial of the number.
    /// </item>
    /// </list>
    ///
    /// <b>Example:</b>
    /// <code>
    /// using Prog.Numbers.Utils.Operations;
    /// fact = FactorialUtils.Factorial(10);
    /// Console.WriteLine(fact);</code>
    ///
    /// <b>Output:</b>
    /// <code>
    /// "3628800"</code>
    /// </summary>
    public static BigInteger Factorial(int num)
    {
        BigInteger fact = 1;

        for (int i = 1; i <= num; i++)
        {
            fact *= i;
        }

        return fact;
    }

    /// <summary> <h1><c>FactorialRandom</c></h1>
    /// <br/><br/>This method is used to find the factorial of a random
    /// number.<br/><br/>
    ///
    /// <b>Return:</b>
    ///
    /// <list type="bullet">
    /// <item>
    /// <c>string</c> - The factorial of the random number.
    /// </item>
    /// </list>
    ///
    /// <b>Example:</b>
    ///
    /// <code>
    /// using Prog.Numbers.Utils.Operations;
    /// fact = FactorialUtils.FactorialRandom();
    /// Console.WriteLine(fact);</code>
    /// </summary>
    public static string FactorialRandom()
    {
        Random rng = new();
        int randomNum = rng.Next(1, 26);

        BigInteger fact = Factorial(randomNum);

        return $"Fact[{randomNum}] = {fact}";
    }
}

File: haskell/packages/numbers-utils/src/Numbers/Utils/Operations/FactorialUtils.hs
module Numbers.Utils.Operations.FactorialUtils (factorial, factorialRandom) where

import System.Random (randomRIO)

-- |
-- # @factorial@
--
-- The function @factorial@ is used to calculate the factorial of a number.
--
-- __Arguments:__
--
-- * @num: Integer@ - The number to calculate the factorial.
--
-- __Return:__
--
-- * @Integer@ - The factorial of the number.
--
-- __Example:__
--
-- @
-- import Numbers.Utils.Operations.FactorialUtils (factorial)
-- let fact = factorial 25
-- putStrLn $ "Fact[25] = " ++ show fact
-- @
--
-- __Output:__
--
-- @
-- "Fact[25] = 15511210043330985984000000"
-- @
factorial :: Integer -> Integer
factorial num = product [1 .. num]

-- |
-- # @factorialRandom@
--
-- The function @factorialRandom@ is used to calculate the factorial of a random
-- number.
--
-- __Return:__
--
-- * @String@ - The factorial of the random number.
--
-- __Example:__
--
-- @
-- import Numbers.Utils.Operations.FactorialUtils (factorialRandom)
-- fact <- factorialRandom
-- putStrLn fact
-- @
--
-- __Output:__
--
-- @
-- "Fact[5] = 120"
-- @
factorialRandom :: IO String
factorialRandom = do
    randomNum <- randomRIO (1, 25 :: Integer)
    let fact = factorial randomNum
    return $ "Fact[" ++ show randomNum ++ "] = " ++ show fact
